schema-manager/src/main.rs

```diff
@@ -0,0 +1,275 @@
+    let zone_uuid = get_zone_uuid()?;
+    sapi.get_zone_config(&zone_uuid)
+}
+
+// Get the boray zone UUID for use in the sapi config request
+fn get_zone_uuid() -> FunResult {
+    run_fun!("/usr/bin/zonename")
+}
+
+// Get a sapi client from the URL in the zone config
+fn init_sapi_client(sapi_address: String, log: Logger) -> Result<SAPI, Error> {
+    Ok(SAPI::new(&sapi_address, 60, log.clone()))
+}
+
+// Iterated through the vnodes returned from electric-boray and create the
+// associated schemas on the shards
+fn parse_vnodes(msg: &FastMessage) -> Result<(), Error> {
+    let v: Vec<Value> = msg.data.d.as_array().unwrap().to_vec();
+    for vnode in v {
+        for v in vnode.as_array().unwrap() {
+            let vn = v.as_str().unwrap();
+            println!("processing vnode: {:#?}", vn);
+            create_bucket_schemas(v.as_str().unwrap())?;
+        }
+    }
+    Ok(())
+}
+
+// Not really used for anything yet
+fn parse_opts<'a, 'b>(app: String) -> ArgMatches<'a> {
+   App::new(app)
+        .version(crate_version!())
+        .about("Tool to manage postgres schemas for boray")
+        .arg(
+            Arg::with_name("fast_args")
+                .help("JSON-encoded arguments for RPC method call")
+                .long("args")
+                .takes_value(true)
+                .required(false),
+        )
+        .get_matches()
+}
+
+// Schema template is stored in boray/schema_templates/schema.in
+fn read_schema_template() -> Result<String, Error> {
+    fs::read_to_string(SCHEMA_STR)
+}
+
+// Admin template is stored in boray/schema_templates/admin.in
+fn read_admin_template() -> Result<String, Error> {
+    fs::read_to_string(ADMIN_STR)
+}
+
+// db template is stored in boray/schema_templates/db.in
+fn read_db_template() -> Result<String, Error> {
+    fs::read_to_string(DB_STR)
+}
+
+// Callback function handed in to the fast::recieve call.
+// NB: this is an asynchronous call back so the arity is prescribed.
+// This means we cannot send in extra bits for later use.
+fn vnode_response_handler(msg: &FastMessage) -> Result<(), Error> {
+    match msg.data.m.name.as_str() {
+        "getvnodes" => {
+            parse_vnodes(msg)?;
+        }
+        _ => println!("Received unrecognized {} response", msg.data.m.name),
+    }
+
+    Ok(())
+}
+
+// Do the deed
+fn run(
+    log: Logger) -> Result<(), Box<dyn std::error::Error>>
+{
+    let sapi_url = get_sapi_url()?;
+    println!("sapi_url:{}", sapi_url);
+    let sapi = init_sapi_client(sapi_url, log.clone())?;
+    let zone_config = get_zone_config(sapi)?;
+    let eb_address = zone_config.metadata.electric_boray;
+    let boray_host = zone_config.metadata.service_name;
+    let boray_config = get_boray_config();
+    let boray_port = boray_config.server.port;
+
+    let fast_arg = [String::from("tcp://"), boray_host, String::from(":"), boray_port.to_string()]
+                    .concat();
+    println!("pnode argument to electric-boray:{}", fast_arg);
```

Comment:
I think we ought to replace the `println` macros with the logging using the `log` instance that gets passed in.

Reply:
I agree -- however I don't see a way to pass it along through the `fast-client::receive` callback to get to the other functions. I did try refactoring the script to be more OO, but that is still a blocking point. I am probably missing something.

The log is needed for passing in to the SAPI client.